Server/GameServer/User.cpp

```diff
@@ -1180,6 +1180,37 @@ void CUser::SendTime()
 }
 
 /**
+* @brief Overloads CUser::SendTime() to change the server time by GM command.
+*/
+void CUser::SendTime(Packet & pkt)
+{
+	if (!this->isGM()) return;
+
+	int16_t year  = pkt.read<int16_t>();
+	int16_t month = pkt.read<int16_t>();
+	int16_t day   = pkt.read<int16_t>();
+	int16_t hour  = pkt.read<int16_t>();
+	int16_t min   = pkt.read<int16_t>();
+
+	if (year  != NULL) g_pMain->m_sYear  = year;
+	if (month != NULL) g_pMain->m_sMonth = month;
+	if (day   != NULL) g_pMain->m_sDate  = day;
+	if (hour  != NULL) g_pMain->m_sHour  = hour;
+	if (min   != NULL) g_pMain->m_sMin   = min;
+
+	if (year == NULL && month == NULL && day == NULL && hour == NULL && min == NULL)
```

Comment:
None of these are nullable types. The only reason this works is because NULL evaluates to 0, but this it not obvious.
This is why we use nullptr over NULL now; this would fail to compile there.

Further, I don't think this is exactly what you wanted.  You want to set the time but only if it's not set to 0... but there are cases where this is intended, e.g. in your sample with time 18:00 -- the minutes portion is 0. Yet since that's 0, it would not update that so the minutes would not be updated.

Reply:
Changed to == 0. 
Yea, that's right regarding the minutes being :00 in certain situations. I had this in mind, but since it will anyway read the packet as 0 when it's empty, there is no point updating stuff in the pointer for CGameServerDlg and for that purpose i included a check there. There is a different issue with the server not resetting the packets for some reason which at the second time you'll use the command, the server will read the old values. I wanted to create an issue for it, but i didn't really had the time to explore this more to add useful details.

Reply:
The problem still stands, though. I'm well aware of how the packet handling works; I wrote it, after all.

The problem here is that you're having to "check" for this, and you're doing so in an unclear manner.

Firstly, why is this even an issue: why are we not providing all of this data at all times?
The official packet structure is as follows:
```
		uint16_t year, month, day, hour, minute;
		pkt >> year >> month >> day >> hour >> minute;
```
That's the structure. All of the data is always provided -- always.

If we're not always providing it in the client, the client is doing it wrong.
To address this, we can always send what the client deems the current time in place of blanks, and hence always send it a value.

Now if this wasn't an official packet (but it is, so we should treat it accordingly), we could otherwise address this in a couple of different ways.
We could:
1. design the packet such that this behaviour is obvious, and explicitly specify what we're changing or,
2. keeping it similar to the above, we could instead check to see if there's more to read from the packet.

Note that the existing checks make no sense, because if at any point something "isn't set", the rest "won't be set" either -- because there won't be anything left in the packet to read from.

The behaviour really doesn't make sense.

But to summarise, since this *is* an official packet, we should make sure that it behaves as it does officially. That is, the client will send _all_ of the date/time data, and the server will handle _all_ of them, always. If we exclude data from the client, it will still supply the current date/time data to the server.

Also note that this date/time isn't actually realtime. It moves at a speed much faster (I don't remember how fast, 6x? 10x? I'd have to check) than realtime, so we also don't want to be getting that mixed up when we do that. :)